### PR TITLE
Added ability to copy a specific version of an app

### PR DESCRIPTION
**Pull request by orangejenny**
The first commit just cherry-picks https://github.com/dimagi/commcare-hq/pull/24928 (and https://github.com/dimagi/commcare-hq/pull/24943) so we can release this functionality separately, as discussed in meeting with QA earlier this week.

[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-666)

@dimagi/product Are y'all okay with releasing this to all users?

Screen Shot 2019-08-15 at 4 16 12 PM

### CODE REVIEW

File: corehq/apps/app_manager/models.py
Comment (orangejenny) [inline corehq/apps/app_manager/models.py]: So I'm not totally in love with this change, since `ApplicationBase` is already 600 lines long, and a VersionedDoc is a relatively clear concept (although the ODK urls don't belong in it...). But I was finding it hard to work with (because VersionedDoc controls copy_of but ApplicationBase controls all the other build-related fields). I'd rather collapse the unused level of hierarchy and instead break ApplicationBase up using composition.

Open to feedback. @proteusvacuum @emord you've also thought about the app manager models hierarchy.

Comment (mkangia) [reply corehq/apps/app_manager/models.py]: I would push for keeping `VersionedDoc`, retaining version related logic in it.

And for build related things to go into application or if possible, in a different class and then that is used to make build related things instead of populating `Application` class. Inheriting `Application` from it would be the easier way out but let's avoid that. I have extracted classes like `HostedCCZUtility` which dealt with creating the CCZ file for `HostedCCZ` and related operations (the build in this case for Application) and then use that to handle operations.

**Review by mkangia — COMMENTED** [review]
not approving yet, would like to see where the conversation for `VersionedDoc` goes from few reviews.

**Comment by orangejenny** [comment]
Note to self: remember to handle feedback from https://github.com/dimagi/commcare-hq/pull/24928#discussion_r307824546

File: corehq/apps/app_manager/models.py
Comment (esoergel) [inline corehq/apps/app_manager/models.py]: I really like the framing of `convert_app_to_build` and vice-versa, but I do notice that `_meta_fields` is used in `export_json` as a list of fields to be excluded.  I think it implements the same concept as `convert_build_to_app`.  If that's true, it'd be good to combine the two (maybe do the `_meta_fields` stuff here in `convert_build_to_app`, say)

**Review by esoergel — APPROVED** [review]

File: corehq/apps/app_manager/views/apps.py
Comment (esoergel) [inline corehq/apps/app_manager/views/apps.py]: Shouldn't `app` here come from `app_id_or_source` instead?

Comment (orangejenny) [reply corehq/apps/app_manager/views/apps.py]: Good catch, this wasn't working for linked apps. Added a few commits to fix, clean up, and add a test.

File: corehq/apps/app_manager/views/apps.py
Comment (esoergel) [inline corehq/apps/app_manager/views/apps.py]: Not sure what the correct approach is, but this needs some rethinking - if `master_build` is `None`, `master_build.domain` will fail above this, and `master_build.master_id` will fail below.

Comment (orangejenny) [reply corehq/apps/app_manager/views/apps.py]: Clearly I'm having trouble getting my head all the way around `copy_app`. I took a fresh look at it and refactored in https://github.com/dimagi/commcare-hq/pull/25003/commits/ad3fb0a32d62a62cec26eb949221f8aeba2a307e with two goals: 1) move as much logic as possible inside `_inner`, rather than having some logic before it and then some inside and 2) make the function take two separate paths (normal vs linked app) that don't share much logic.

**Review by esoergel — APPROVED** [review]
Nice, thanks - I also had trouble wrapping my head around this functionality

File: corehq/apps/app_manager/models.py
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: I don't think these are quite the same pieces of functionality, given that `VersionedDoc.make_build` excludes a similar but not identical list of properties: https://github.com/dimagi/commcare-hq/blob/dcaeefa1dd579d43b1eabe6795216c555f3d2eb5/corehq/apps/app_manager/models.py#L3912

File: corehq/apps/app_manager/models.py
Comment (orangejenny) [reply corehq/apps/app_manager/models.py]: Looked at this a bit more, and I still like the idea of making VersionedDoc a separate utility class, but doing that is more complicated than I expected. VersionedDoc isn't a clean abstraction, given its use of application-specific urls (short_odk_url, etc) and the `app_manager/applications` couch view.

I don't think it's worth the effort to clean up VersionedDoc enough to make it truly generic and pull it out of the Application hierarchy, and I still think the extra layer of hierarchy (e.g., having two implementations of `make_build`) adds more in complexity than it's worth. I'm inclined to leave this code as it is in this PR, removing VersionedDoc altogether.

File: corehq/apps/app_manager/models.py
Comment (esoergel) [reply corehq/apps/app_manager/models.py]: Yeah, I see that as well.  Seems like there's some overlap between `convert_build_to_app`, `export_json`, and `from_source`, at least conceptually, but I do see there are differences too.  Not sure whether the differences are intentional/desirable or accidental.  I'll leave it to your judgement.

**Comment by orangejenny** [comment]
Marking don't merge again until I update the labeling per discussion with product.

**Comment by esoergel** [comment]
:tada:

**Pull request merged**
